File: src/lib/api/git/status.ts
// Working-tree status and every diff read — file, commit, range, selection, and
// comparison — plus blame and file history. Mirrors `commands/status.rs`.

import { invoke } from "@/lib/api/invoke";
import { parse } from "@/lib/api/validate";
import { fileDiffSchema, workingChangesSchema } from "@/lib/api/schemas";
import type {
  BinaryBlob,
  CompareResult,
  FileBlame,
  FileChange,
  FileDiff,
  FileHistoryPage,
  WorkingChanges,
} from "./types";

export const statusApi = {
  workingChanges: async (path: string): Promise<WorkingChanges> =>
    // The schema defaults `conflicted` to [] (the long-standing defensive
    // contract, so every consumer can rely on the field) and rejects any other
    // shape drift with a clear IpcValidationError.
    parse(workingChangesSchema, await invoke("working_changes", { path }), "working_changes"),

  /** Diff for a working-tree file. `staged` true → index vs HEAD; false → worktree vs index.
   * `full` bypasses the backend line cap (for an explicit "show full diff"). */
  fileDiff: async (path: string, file: string, staged: boolean, full?: boolean): Promise<FileDiff> =>
    parse(fileDiffSchema, await invoke("file_diff", { path, file, staged, full: full ?? null }), "file_diff"),

  /** Changed files in a commit (vs its first parent). */
  commitFiles: (path: string, oid: string) =>
    invoke<FileChange[]>("commit_files", { path, oid }),

  /** Read a binary blob's bytes (base64) for an inline preview. Pass `oid` for a
   * committed/staged blob; pass `file` (repo-relative, `oid` omitted) to read the
   * working-tree copy — the side an unstaged diff leaves without a blob oid. */
  readBinaryBlob: (
    path: string,
    source: { oid?: string | null; file?: string | null },
    maxBytes?: number,
  ) =>
    invoke<BinaryBlob>("read_binary_blob", {
      path,
      oid: source.oid ?? null,
      file: source.file ?? null,
      maxBytes: maxBytes ?? null,
    }),

  /** Diff for one file within a commit (vs its first parent). `full` bypasses
   * the backend line cap (for an explicit "show full diff"). */
  commitFileDiff: async (path: string, oid: string, file: string, full?: boolean): Promise<FileDiff> =>
    parse(
      fileDiffSchema,
      await invoke("commit_file_diff", { path, oid, file, full: full ?? null }),
      "commit_file_diff",
    ),

  /** Changed files across a range base..head (either side accepts any
   * commit-ish: a SHA, "HEAD", a branch). */
  diffRange: (path: string, base: string, head: string) =>
    invoke<FileChange[]>("diff_range", { path, base, head }),

  /** Diff for one file across a range base..head. `full` bypasses the backend
   * line cap (for an explicit "show full diff"). */
  diffRangeFile: async (
    path: string,
    base: string,
    head: string,
    file: string,
    full?: boolean,
  ): Promise<FileDiff> =>
    parse(
      fileDiffSchema,
      await invoke("diff_range_file", { path, base, head, file, full: full ?? null }),
      "diff_range_file",
    ),

  /** Merged ("union") changed files across a multi-commit selection (GL-69): the
   * net change per file across `oids` (in any order), with status + counts. For
   * each file the net is computed from its state before the earliest selected
   * commit that touches it to its state after the latest one. */
  selectionDiff: (path: string, oids: string[]) =>
    invoke<FileChange[]>("selection_diff", { path, oids }),

  /** Merged diff for one file across a multi-commit selection (see
   * {@link selectionDiff}). `full` bypasses the backend line cap. */
  selectionDiffFile: async (
    path: string,
    oids: string[],
    file: string,
    full?: boolean,
  ): Promise<FileDiff> =>
    parse(
      fileDiffSchema,
      await invoke("selection_diff_file", { path, oids, file, full: full ?? null }),
      "selection_diff_file",
    ),

  /** Bounded newest-first history for a repository-relative file path. */
  fileHistory: (path: string, file: string, offset?: number, limit?: number) =>
    invoke<FileHistoryPage>("file_history", {
      path,
      file,
      offset: offset ?? null,
      limit: limit ?? null,
    }),

  /** Line-level attribution for a text file at a revision or the working tree. */
  fileBlame: (path: string, file: string, revision?: string | null, limit?: number) =>
    invoke<FileBlame>("file_blame", {
      path,
      file,
      revision: revision ?? null,
      limit: limit ?? null,
    }),

  /** Changed files plus totals for a `base..head` comparison. `head = null`
   * compares `base` against the working tree. */
  compareRefs: (path: string, base: string, head?: string | null) =>
    invoke<CompareResult>("compare_refs", { path, base, head: head ?? null }),

  /** Full diff for one file within a comparison (see [`compareRefs`]). */
  compareFileDiff: async (
    path: string,
    base: string,
    head: string | null,
    file: string,
    full?: boolean,
  ): Promise<FileDiff> =>
    parse(
      fileDiffSchema,
      await invoke("compare_file_diff", { path, base, head: head ?? null, file, full: full ?? null }),
      "compare_file_diff",
    ),
};

File: src-tauri/src/commands/status.rs
//! Working-tree status and every diff read (commit, range, selection, compare, history, blame).

use super::{blocking, CommandError};
use crate::git;
use crate::git::types::{
    BinaryBlob, CompareResult, FileBlame, FileChange, FileDiff, FileHistoryPage, WorkingChanges,
};

// Every read here scales with the working tree, the commit, or the range it
// inspects — a status walk over tens of thousands of files, a tree diff, a
// blob diff. Like `commit_graph`, all of them run on the blocking pool so the
// webview keeps repainting while they work (`ipc/commands` spec); the non-Send
// Repository is opened and dropped inside each closure.

#[tauri::command]
pub async fn working_changes(path: String) -> Result<WorkingChanges, CommandError> {
    blocking(move || git::status::working_changes(&path).map_err(|e| e.to_string())).await
}

#[tauri::command]
pub async fn file_diff(
    path: String,
    file: String,
    staged: bool,
    full: Option<bool>,
) -> Result<FileDiff, CommandError> {
    blocking(move || {
        git::status::file_diff(&path, &file, staged, full.unwrap_or(false))
            .map_err(|e| e.to_string())
    })
    .await
}

#[tauri::command]
pub async fn commit_files(path: String, oid: String) -> Result<Vec<FileChange>, CommandError> {
    blocking(move || git::status::commit_files(&path, &oid).map_err(|e| e.to_string())).await
}

/// Read a binary blob's bytes (base64) for an inline preview. `oid` selects a
/// committed/staged blob; omit it (with `file` set) to read the working-tree
/// file by path — the side libgit2 leaves without a blob oid in an unstaged diff.
///
/// Reads up to a few MiB off disk/ODB and base64-encodes it, so it runs on the
/// blocking pool (like `commit_graph`) to keep the webview thread responsive when
/// several image panes load at once.
#[tauri::command]
pub async fn read_binary_blob(
    path: String,
    oid: Option<String>,
    file: Option<String>,
    max_bytes: Option<u64>,
) -> Result<BinaryBlob, CommandError> {
    blocking(move || {
        git::status::read_binary_blob(&path, oid.as_deref(), file.as_deref(), max_bytes)
            .map_err(|e| e.to_string())
    })
    .await
}

#[tauri::command]
pub async fn commit_file_diff(
    path: String,
    oid: String,
    file: String,
    full: Option<bool>,
) -> Result<FileDiff, CommandError> {
    blocking(move || {
        git::status::commit_file_diff(&path, &oid, &file, full.unwrap_or(false))
            .map_err(|e| e.to_string())
    })
    .await
}

#[tauri::command]
pub async fn diff_range(
    path: String,
    base: String,
    head: String,
) -> Result<Vec<FileChange>, CommandError> {
    blocking(move || git::status::diff_range(&path, &base, &head).map_err(|e| e.to_string())).await
}

#[tauri::command]
pub async fn diff_range_file(
    path: String,
    base: String,
    head: String,
    file: String,
    full: Option<bool>,
) -> Result<FileDiff, CommandError> {
    blocking(move || {
        git::status::diff_range_file(&path, &base, &head, &file, full.unwrap_or(false))
            .map_err(|e| e.to_string())
    })
    .await
}

// The heaviest of the reads: a multi-thousand-commit history walk (each step
// diffing one file), blame over a long file, or a full-tree comparison.
#[tauri::command]
pub async fn file_history(
    path: String,
    file: String,
    offset: Option<usize>,
    limit: Option<usize>,
) -> Result<FileHistoryPage, CommandError> {
    blocking(move || {
        git::status::file_history(&path, &file, offset, limit).map_err(|e| e.to_string())
    })
    .await
}

#[tauri::command]
pub async fn file_blame(
    path: String,
    file: String,
    revision: Option<String>,
    limit: Option<usize>,
) -> Result<FileBlame, CommandError> {
    blocking(move || {
        git::status::file_blame(&path, &file, revision, limit).map_err(|e| e.to_string())
    })
    .await
}

#[tauri::command]
pub async fn compare_refs(
    path: String,
    base: String,
    head: Option<String>,
) -> Result<CompareResult, CommandError> {
    blocking(move || {
        git::status::compare_refs(&path, &base, head.as_deref()).map_err(|e| e.to_string())
    })
    .await
}

#[tauri::command]
pub async fn compare_file_diff(
    path: String,
    base: String,
    head: Option<String>,
    file: String,
    full: Option<bool>,
) -> Result<FileDiff, CommandError> {
    blocking(move || {
        git::status::compare_file_diff(&path, &base, head.as_deref(), &file, full.unwrap_or(false))
            .map_err(|e| e.to_string())
    })
    .await
}

/// Merged ("union") diff across a multi-commit selection (GL-69): the net change
/// per file across an arbitrary `oids` set. Walks one tree diff per selected
/// commit, so it runs on the blocking pool like the other range reads.
#[tauri::command]
pub async fn selection_diff(
    path: String,
    oids: Vec<String>,
) -> Result<Vec<FileChange>, CommandError> {
    blocking(move || git::status::selection_diff(&path, &oids).map_err(|e| e.to_string())).await
}

#[tauri::command]
pub async fn selection_diff_file(
    path: String,
    oids: Vec<String>,
    file: String,
    full: Option<bool>,
) -> Result<FileDiff, CommandError> {
    blocking(move || {
        git::status::selection_diff_file(&path, &oids, &file, full.unwrap_or(false))
            .map_err(|e| e.to_string())
    })
    .await
}
